batcher/aligned-sdk/src/communication/messaging.rs

```
@@ -0,0 +1,213 @@
+};
+
+pub async fn send_messages(
+    ws_write: Arc<Mutex<SplitSink<WebSocketStream<MaybeTlsStream<TcpStream>>, Message>>>,
+    verification_data: &[VerificationData],
+    wallet: Wallet<SigningKey>,
+) -> Result<Vec<VerificationData>, SubmitError> {
+    let mut sent_verification_data = Vec::new();
+
+    let mut ws_write = ws_write.lock().await;
+
+    for verification_data in verification_data.iter() {
+        let msg = ClientMessage::new(verification_data.clone(), wallet.clone()).await;
+        let msg_str = serde_json::to_string(&msg).map_err(SubmitError::SerdeError)?;
+        ws_write
+            .send(Message::Text(msg_str.clone()))
+            .await
+            .map_err(SubmitError::ConnectionError)?;
+        sent_verification_data.push(verification_data.clone());
+        debug!("Message sent...");
+    }
+
+    Ok(sent_verification_data)
+}
+
+pub async fn receive(
+    ws_read: SplitStream<WebSocketStream<MaybeTlsStream<TcpStream>>>,
+    ws_write: Arc<Mutex<SplitSink<WebSocketStream<MaybeTlsStream<TcpStream>>, Message>>>,
+    total_messages: usize,
+    num_responses: Arc<Mutex<usize>>,
+    verification_data_commitments_rev: &mut Vec<VerificationDataCommitment>,
+) -> Result<Option<Vec<AlignedVerificationData>>, SubmitError> {
+    // Responses are filtered to only admit binary or close messages.
+    let mut response_stream =
+        ws_read.try_filter(|msg| future::ready(msg.is_binary() || msg.is_close()));
+
+    let mut aligned_verification_data: Vec<AlignedVerificationData> = Vec::new();
+
+    while let Some(Ok(msg)) = response_stream.next().await {
+        if let Message::Close(close_frame) = msg {
+            if let Some(close_msg) = close_frame {
+                error!("Connection was closed before receiving all messages. Reason: {}. Try submitting your proof again", close_msg.to_owned());
+                ws_write.lock().await.close().await?;
+                return Ok(None);
```

This should return an error, the user must know that the connection was closed before receiving all messages

---

Solved in [cef4afb](https://github.com/yetanotherco/aligned_layer/pull/583/commits/cef4afb1e728da34f45728b42d6780e29a862585)